File: core/src/domain/purchase/purchase_invoice.py
from dataclasses import dataclass

from src.domain.purchase.exception import InvalidPurchaseInvoiceNumberError


@dataclass
class PurchaseInvoice:
    """Represents a purchase invoice."""
    number: str = None
    amount: float = None
    taxes: float = None
    total: float = None

    @classmethod
    def create(
            cls,
            number: str,
            amount: float,
            taxes: float,
            total: float,
    ) -> "PurchaseInvoice":
        """
        Factory method to create a new PurchaseInvoice instance.
        """
        cls.check(number)

        return cls(number, amount, taxes, total)

    @staticmethod
    def check(number: str) -> None:
        """
        Validates the purchase invoice number.
        """
        if not number or not number.strip():
            raise InvalidPurchaseInvoiceNumberError()

File: editor/src/app/purchase/model/purchase-invoice.ts
export class PurchaseInvoice {
  constructor(
    public number: string,
    public amount: number,
    public taxes: number,
    public total: number,
  ) {}

  public static from(item: PurchaseInvoice): PurchaseInvoice {
    return new PurchaseInvoice(
      item.number,
      item.amount,
      item.taxes,
      item.total,
    );
  }
}
